numpy/core/src/multiarray/refcount.c

```diff
@@ -51,6 +51,21 @@ PyArray_Item_INCREF(char *data, PyArray_Descr *descr)
             PyArray_Item_INCREF(data + offset, new);
         }
     }
+    else if (PyDataType_HASSUBARRAY(descr)) {
```

Comment:
I'm confused: shouldn't this happen only for a subarray of object dtype?

Reply:
I think it could be of `O,f` type for example

Reply:
Of course, if I try to do that I get segmentation faults, so it is unlikely anyone ever did that ;).

Reply:
no, we might end up with a subarray which is a structure with an object field, eg `np.dtype([('a', 'i4'), ('b', [('ba', 'O'), ('bb', 'i4')])])`. I forgot about that above too.

Reply:
Hmm, what the segfault example?

Reply:
Just create an array with your dtype and then try to access the "b" field.

Reply:
Seems a bit strange if I am honest, printing that part crashes with invalid object cast, but I am not quite sure what goes wrong. The dtype and pointers look fine on first sight.

Reply:
So the problem is, that it fills the object parts incorrectly with 0s, there is an offset to it or so, which causes most of them to be wrong.

Reply:
I can't get it to segfault, I tried py2/py3, 1.15.4, master, this PR:
```python
>>> dt = np.dtype([('a', 'i4'), ('b', [('ba', 'O'), ('bb', 'f4')], (2,3))])
>>> a = np.ones(3, dt)
>>> a
array([(1, [[(1, 1.), (1, 1.), (1, 1.)], [(1, 1.), (1, 1.), (1, 1.)]]),
       (1, [[(1, 1.), (1, 1.), (1, 1.)], [(1, 1.), (1, 1.), (1, 1.)]]),
       (1, [[(1, 1.), (1, 1.), (1, 1.)], [(1, 1.), (1, 1.), (1, 1.)]])],
      dtype=[('a', '<i4'), ('b', [('ba', 'O'), ('bb', '<f4')], (2, 3))])
>>> a['b']
array([[[(1, 1.), (1, 1.), (1, 1.)],
        [(1, 1.), (1, 1.), (1, 1.)]],

       [[(1, 1.), (1, 1.), (1, 1.)],
        [(1, 1.), (1, 1.), (1, 1.)]],

       [[(1, 1.), (1, 1.), (1, 1.)],
        [(1, 1.), (1, 1.), (1, 1.)]]], dtype=[('ba', 'O'), ('bb', '<f4')])
>>> a['b'] = (5,4)
>>> a
array([(1, [[(5, 4.), (5, 4.), (5, 4.)], [(5, 4.), (5, 4.), (5, 4.)]]),
       (1, [[(5, 4.), (5, 4.), (5, 4.)], [(5, 4.), (5, 4.), (5, 4.)]]),
       (1, [[(5, 4.), (5, 4.), (5, 4.)], [(5, 4.), (5, 4.), (5, 4.)]])],
      dtype=[('a', '<i4'), ('b', [('ba', 'O'), ('bb', '<f4')], (2, 3))])
```

Reply:
Somewhere the zero/None filling must be not accounting for the second type. But maybe that is exactly the code snippet you pasted above that is at fault....

Reply:
rofl, the ones code works... `zeros` and `empty` do not.

Reply:
I think it's because  other methods in `refcoun.c` are also missing subarray clauses. `np.empty` and `np.zeros` go through `PyArray_FillObjectArray` and `_fillobject` in that file, and the latter ignores subarrays. `np.ones` is saved because it is instead implemented as `a = np.empty(x, dt); a[:] = 1`.

I suggest we open a new issue for this and leave it alone in this PR.

Reply:
That code is a bit strange, but I guess it should be fine to give it the same type of handling as the other ones.

Reply:
Forced push a fix, added fast paths to these two code blocks as well. these definitely should get a bit thorough testing though. Not that it is all that much a few example dtypes such as the one above including refcount checks.

Reply:
Hmm, maybe: When we talk about "float"s we meant python float objects, not actual float/numpy floats. You can replace float with `object`.

Reply:
Hmmmpf, trying to add a test, it seems like the assignment loop also is buggy when it comes to reference counts.

Reply:
So, I can "patch" it over with a hack given in the diff below. The general thing is that it goes to a buffered path and then just initializes the whole buffer with 0s. The buffer itself has no refcounting, the function from the source will thus only ever `XDECREF(NULL)`, while the input just gets overwritten.

I would think there must be a way to go to some slower path which is safe, but not sure. Maybe a solution somewhat like this is actually right...

<details>

```diff
diff --git a/numpy/core/src/multiarray/dtype_transfer.c b/numpy/core/src/multiarray/dtype_transfer.c
index 2b29d4f8c..d34699312 100644
--- a/numpy/core/src/multiarray/dtype_transfer.c
+++ b/numpy/core/src/multiarray/dtype_transfer.c
@@ -464,6 +464,64 @@ _strided_to_strided_contig_align_wrap_init_dest(char *dst, npy_intp dst_stride,
     }
 }
 
+
+
+static void
+_strided_to_strided_contig_align_wrap_init_dest_refcnt(char *dst, npy_intp dst_stride,
+                        char *src, npy_intp src_stride,
+                        npy_intp N, npy_intp src_itemsize,
+                        NpyAuxData *data)
+{
+    _align_wrap_data *d = (_align_wrap_data *)data;
+    PyArray_StridedUnaryOp *wrapped = d->wrapped,
+            *tobuffer = d->tobuffer,
+            *frombuffer = d->frombuffer;
+    npy_intp inner_src_itemsize = d->src_itemsize,
+             dst_itemsize = d->dst_itemsize;
+    NpyAuxData *wrappeddata = d->wrappeddata,
+            *todata = d->todata,
+            *fromdata = d->fromdata;
+    char *bufferin = d->bufferin, *bufferout = d->bufferout;
+    PyArray_StridedUnaryOp *_strided_to_strided = PyArray_GetStridedCopyFn(
+            0, dst_stride, sizeof(PyObject *), sizeof(PyObject *));
+
+    for(;;) {
+        if (N > NPY_LOWLEVEL_BUFFER_BLOCKSIZE) {
+            tobuffer(bufferin, inner_src_itemsize, src, src_stride,
+                                    NPY_LOWLEVEL_BUFFER_BLOCKSIZE,
+                                    src_itemsize, todata);
+            _strided_to_strided(
+                    bufferout, sizeof(PyObject *),
+                    dst, dst_stride, NPY_LOWLEVEL_BUFFER_BLOCKSIZE,
+                    sizeof(PyObject *), NULL);
+            wrapped(bufferout, dst_itemsize, bufferin, inner_src_itemsize,
+                                    NPY_LOWLEVEL_BUFFER_BLOCKSIZE,
+                                    inner_src_itemsize, wrappeddata);
+            frombuffer(dst, dst_stride, bufferout, dst_itemsize,
+                                    NPY_LOWLEVEL_BUFFER_BLOCKSIZE,
+                                    dst_itemsize, fromdata);
+            N -= NPY_LOWLEVEL_BUFFER_BLOCKSIZE;
+            src += NPY_LOWLEVEL_BUFFER_BLOCKSIZE*src_stride;
+            dst += NPY_LOWLEVEL_BUFFER_BLOCKSIZE*dst_stride;
+        }
+        else {
+            tobuffer(bufferin, inner_src_itemsize, src, src_stride, N,
+                                            src_itemsize, todata);
+            _strided_to_strided(
+                    bufferout, sizeof(PyObject *),
+                    dst, dst_stride, N,
+                    sizeof(PyObject *), NULL);
+            wrapped(bufferout, dst_itemsize, bufferin, inner_src_itemsize, N,
+                                            inner_src_itemsize, wrappeddata);
+            frombuffer(dst, dst_stride, bufferout, dst_itemsize, N,
+                                            dst_itemsize, fromdata);
+            return;
+        }
+    }
+}
+
+
+
 /*
  * Wraps an aligned contig to contig transfer function between either
  * copies or byte swaps to temporary buffers.
@@ -485,7 +543,7 @@ wrap_aligned_contig_transfer_function(
             PyArray_StridedUnaryOp *tobuffer, NpyAuxData *todata,
             PyArray_StridedUnaryOp *frombuffer, NpyAuxData *fromdata,
             PyArray_StridedUnaryOp *wrapped, NpyAuxData *wrappeddata,
-            int init_dest,
+            int init_dest, int refcnt_dest,
             PyArray_StridedUnaryOp **out_stransfer,
             NpyAuxData **out_transferdata)
 {
@@ -520,7 +578,10 @@ wrap_aligned_contig_transfer_function(
                 NPY_LOWLEVEL_BUFFER_BLOCKSIZE*src_itemsize;
 
     /* Set the function and data */
-    if (init_dest) {
+    if (refcnt_dest) {
+        *out_stransfer = &_strided_to_strided_contig_align_wrap_init_dest_refcnt;
+    }
+    else if (init_dest) {
         *out_stransfer = &_strided_to_strided_contig_align_wrap_init_dest;
     }
     else {
@@ -1170,6 +1231,7 @@ get_datetime_to_unicode_transfer_function(int aligned,
                         frombuffer, fromdata,
                         caststransfer, castdata,
                         PyDataType_FLAGCHK(str_dtype, NPY_NEEDS_INIT),
+                        0,
                         out_stransfer, out_transferdata) != NPY_SUCCEED) {
         NPY_AUXDATA_FREE(castdata);
         NPY_AUXDATA_FREE(todata);
@@ -1292,6 +1354,7 @@ get_unicode_to_datetime_transfer_function(int aligned,
                         frombuffer, fromdata,
                         caststransfer, castdata,
                         PyDataType_FLAGCHK(dst_dtype, NPY_NEEDS_INIT),
+                        0,
                         out_stransfer, out_transferdata) != NPY_SUCCEED) {
         Py_DECREF(str_dtype);
         NPY_AUXDATA_FREE(castdata);
@@ -1594,6 +1657,7 @@ get_cast_transfer_function(int aligned,
                             frombuffer, fromdata,
                             caststransfer, castdata,
                             PyDataType_FLAGCHK(dst_dtype, NPY_NEEDS_INIT),
+                            dst_dtype->type_num == NPY_OBJECT,
                             out_stransfer, out_transferdata) != NPY_SUCCEED) {
             NPY_AUXDATA_FREE(castdata);
             NPY_AUXDATA_FREE(todata);
```

</details>

<par><par>
The test case is (probably can be prettied up a bit, plus I am not quite sure if the refcounts might be a bit flaky at times):
<details>

```python
def test_structured_with_object():
    # Test that object creation and access works for complex dtypes
    # with subarrays.
    dt1 = np.dtype([('a', 'i4'),
                    ('b', [('ba', 'O'), ('bb', 'i1')], (2, 3))])
    dt2 = np.dtype([('a', 'i4'),
                    ('b', [('ba', 'O'), ('bb', 'O')], (2, 3))])
    dt3 = np.dtype([('a', 'i4'), ('b', 'O', (2, 3))])

    # create patterns to fill array manually:
    singleton = object()
    p1 = (0, [[(singleton, 0)] * 3] * 2)
    p2 = (0, [[(singleton, singleton)] * 3] * 2)
    p3 = (0, [[singleton] * 3] * 2)

    one = 1
    zero = 0
    for dt, count, pat in zip([dt1, dt2, dt3], [6, 12, 6], [p1, p2, p3]):
        # Test assumes that 0, 1, and None are singletons.
        for func, obj in zip([np.empty, np.ones, np.zeros], [None, one, zero]):
            if obj is None:
                # None test is flaky:
                continue
            gc.collect()
            before = sys.getrefcount(obj)
            print(func, 3, dt)
            arr = func(3, dt)
            
            now = sys.getrefcount(obj)
            assert now - before == count * 3
            del arr
            now = sys.getrefcount(obj)
            assert now == before

        gc.collect()
        before = sys.getrefcount(singleton)
        arr = np.array([pat] * 3, dt)
        assert sys.getrefcount(singleton) - before == count * 3
        # Fill with `1` and check that it was replaced correctly:
        before2 = sys.getrefcount(one)
        arr[...] = one
        #wtf =  sys.getrefcount(singleton), before
        #wtf2 = sys.getrefcount(one) - before2, count * 3
        after2 = sys.getrefcount(one)
        assert after2 - before2 == count * 3
        del arr
        gc.collect()
        assert sys.getrefcount(one) == before2
        assert sys.getrefcount(singleton) == before
```

</details>